TypeError raised in _execute_task

I've run into an issue that may or may not be a bug. I'm using `dask.array.map_blocks` to execute a custom processing function. Dask is raising a `TypeError` in `_execute_task` and I've had quite a difficult time sorting out the reason for the error. @mrocklin was thinking it may be that I "have constructed a graph by hand and that that graph is not consistent. Some key is listed as a dependency but is not actually in the task graph". 

My dask arrays are originating from within Xarray so I'm thinking part of the task tree is being constructed incorrectly there. In fact, I can only reproduce the error if pull my dask arrays from xarray objects (if I strip out the xarray layer, I can run my function without any problem). If true, I'm mostly asking for a more informative error message in this case.

The basic gist of what I'm trying to do is:

```Python
new = da.map_blocks(my_func, a, b, c, chunks=a.chunks, dtype=a.dtype).compute()
```
where: 
```Python
def my_func(a, b, c):
    new = np.empty_like(a)
    # does some things using b and c
    return new
```

I end up generating a task tree that looks like:

![mydask](https://cloud.githubusercontent.com/assets/2443309/25717786/d4a42b5e-30c0-11e7-974c-dddb5d32abf3.png)

What I'm looking for here is a better way to tease out why dask is failing and how I (or Xarray) is creating an invalid task tree.

**Version info**
OS: Mac OS 10.12.4
Python
Xarray: 0.9.5
Dask: 0.14.1

**Traceback**

```Python-traceback
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/xarray/core/dataarray.py:590: in compute
    return new.load()
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/xarray/core/dataarray.py:573: in load
    ds = self._to_temp_dataset().load()
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/xarray/core/dataset.py:469: in load
    evaluated_data = da.compute(*lazy_data.values())
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/base.py:202: in compute
    results = get(dsk, keys, **kwargs)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:544: in get_sync
    raise_on_exception=True, **kwargs)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:487: in get_async
    fire_task()
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:483: in fire_task
    callback=queue.put)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:532: in apply_sync
    res = func(*args, **kwds)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:266: in execute_task
    result = _execute_task(task, data)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:246: in _execute_task
    args2 = [_execute_task(a, cache) for a in args]
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:246: in <listcomp>
    args2 = [_execute_task(a, cache) for a in args]
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:246: in _execute_task
    args2 = [_execute_task(a, cache) for a in args]
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:246: in <listcomp>
    args2 = [_execute_task(a, cache) for a in args]
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:247: in _execute_task
    return func(*args2)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/array/core.py:81: in getarray_inline
    return getarray(a, b, lock=lock)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/array/core.py:64: in getarray
    c = np.asarray(c)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/numpy/core/numeric.py:531: in asarray
    return array(a, dtype, copy=False, order=order)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/xarray/core/common.py:94: in __array__
    return np.asarray(self.values, dtype=dtype)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/xarray/core/dataarray.py:401: in values
    return self.variable.values
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/xarray/core/variable.py:308: in values
    return _as_array_or_item(self._data)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/xarray/core/variable.py:184: in _as_array_or_item
    data = np.asarray(data)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/numpy/core/numeric.py:531: in asarray
    return array(a, dtype, copy=False, order=order)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/array/core.py:1056: in __array__
    x = self.compute()
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/base.py:95: in compute
    (result,) = compute(self, traverse=False, **kwargs)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/base.py:202: in compute
    results = get(dsk, keys, **kwargs)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:544: in get_sync
    raise_on_exception=True, **kwargs)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:487: in get_async
    fire_task()
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:483: in fire_task
    callback=queue.put)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:532: in apply_sync
    res = func(*args, **kwds)
../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:266: in execute_task
    result = _execute_task(task, data)

-----

arg = (<built-in function getitem>, ('data_to_match_array', 0, 0, 0), (0, slice(None, None, None), slice(None, None, None))), cache = {}, dsk = None

    def _execute_task(arg, cache, dsk=None):
        """ Do the actual work of collecting data and executing a function
    
        Examples
        --------
    
        >>> cache = {'x': 1, 'y': 2}
    
        Compute tasks against a cache
        >>> _execute_task((add, 'x', 1), cache)  # Compute task in naive manner
        2
        >>> _execute_task((add, (inc, 'x'), 1), cache)  # Support nested computation
        3
    
        Also grab data from cache
        >>> _execute_task('x', cache)
        1
    
        Support nested lists
        >>> list(_execute_task(['x', 'y'], cache))
        [1, 2]
    
        >>> list(map(list, _execute_task([['x', 'y'], ['y', 'x']], cache)))
        [[1, 2], [2, 1]]
    
        >>> _execute_task('foo', cache)  # Passes through on non-keys
        'foo'
        """
        if isinstance(arg, list):
            return [_execute_task(a, cache) for a in arg]
        elif istask(arg):
            func, args = arg[0], arg[1:]
            args2 = [_execute_task(a, cache) for a in args]
>           return func(*args2)
E           TypeError: tuple indices must be integers or slices, not tuple

../../../../anaconda/envs/storylines/lib/python3.6/site-packages/dask/async.py:247: TypeError
```

cc @shoyer

--- a/dask/async.py
+++ b/dask/async.py
@@ -116,9 +116,8 @@ from __future__ import absolute_import, division, print_function
 
 import os
 import sys
-import traceback
 
-from .compatibility import Queue, Empty
+from .compatibility import Queue, Empty, reraise
 from .core import (istask, flatten, reverse_dict, get_dependencies, ishashable,
                    has_tasks)
 from .context import _globals
@@ -278,7 +277,7 @@ def _execute_task(arg, cache, dsk=None):
         return arg
 
 
-def execute_task(key, task_info, dumps, loads, get_id, raise_on_exception=False):
+def execute_task(key, task_info, dumps, loads, get_id, pack_exception):
     """
     Compute task and handle all administration
 
@@ -290,21 +289,12 @@ def execute_task(key, task_info, dumps, loads, get_id, raise_on_exception=False)
         task, data = loads(task_info)
         result = _execute_task(task, data)
         id = get_id()
-        result = dumps((result, None, id))
+        result = dumps((result, id))
+        failed = False
     except BaseException as e:
-        if raise_on_exception:
-            raise
-        exc_type, exc_value, exc_traceback = sys.exc_info()
-        tb = ''.join(traceback.format_tb(exc_traceback))
-        try:
-            result = dumps((e, tb, None))
-        except BaseException as e:
-            if raise_on_exception:
-                raise
-            exc_type, exc_value, exc_traceback = sys.exc_info()
-            tb = ''.join(traceback.format_tb(exc_traceback))
-            result = dumps((e, tb, None))
-    return key, result
+        result = pack_exception(e, dumps)
+        failed = True
+    return key, result, failed
 
 
 def release_data(key, state, delete=True):
@@ -380,6 +370,10 @@ def default_get_id():
     return None
 
 
+def default_pack_exception(e, dumps):
+    raise
+
+
 def identity(x):
     """ Identity function. Returns x.
 
@@ -409,10 +403,9 @@ The main function of the scheduler.  Get is the main entry point.
 
 
 def get_async(apply_async, num_workers, dsk, result, cache=None,
-              get_id=default_get_id, raise_on_exception=False,
-              rerun_exceptions_locally=None, callbacks=None,
-              dumps=identity, loads=identity,
-              **kwargs):
+              get_id=default_get_id, rerun_exceptions_locally=None,
+              pack_exception=default_pack_exception, raise_exception=reraise,
+              callbacks=None, dumps=identity, loads=identity, **kwargs):
     """ Asynchronous get function
 
     This is a general version of various asynchronous schedulers for dask.  It
@@ -422,12 +415,11 @@ def get_async(apply_async, num_workers, dsk, result, cache=None,
 
     Parameters
     ----------
-
     apply_async : function
         Asynchronous apply function as found on Pool or ThreadPool
     num_workers : int
         The number of active tasks we should have at any one time
-    dsk: dict
+    dsk : dict
         A dask dictionary specifying a workflow
     result : key or list of keys
         Keys corresponding to desired data
@@ -439,6 +431,12 @@ def get_async(apply_async, num_workers, dsk, result, cache=None,
     rerun_exceptions_locally : bool, optional
         Whether to rerun failing tasks in local process to enable debugging
         (False by default)
+    pack_exception : callable, optional
+        Function to take an exception and ``dumps`` method, and return a
+        serialized tuple of ``(exception, traceback)`` to send back to the
+        scheduler. Default is to just raise the exception.
+    raise_exception : callable, optional
+        Function that takes an exception and a traceback, and raises an error.
     dumps: callable, optional
         Function to serialize task data and results to communicate between
         worker and parent.  Defaults to identity.
@@ -451,7 +449,6 @@ def get_async(apply_async, num_workers, dsk, result, cache=None,
 
     See Also
     --------
-
     threaded.get
     """
     queue = Queue()
@@ -504,7 +501,7 @@ def get_async(apply_async, num_workers, dsk, result, cache=None,
             # Submit
             apply_async(execute_task,
                         args=(key, dumps((dsk[key], data)),
-                              dumps, loads, get_id, raise_on_exception),
+                              dumps, loads, get_id, pack_exception),
                         callback=queue.put)
 
         # Seed initial tasks into the thread pool
@@ -513,16 +510,17 @@ def get_async(apply_async, num_workers, dsk, result, cache=None,
 
         # Main loop, wait on tasks to finish, insert new ones
         while state['waiting'] or state['ready'] or state['running']:
-            key, res_info = queue_get(queue)
-            res, tb, worker_id = loads(res_info)
-            if isinstance(res, BaseException):
+            key, res_info, failed = queue_get(queue)
+            if failed:
+                exc, tb = loads(res_info)
                 if rerun_exceptions_locally:
                     data = dict((dep, state['cache'][dep])
                                 for dep in get_dependencies(dsk, key))
                     task = dsk[key]
                     _execute_task(task, data)  # Re-execute locally
                 else:
-                    raise(remote_exception(res, tb))
+                    raise_exception(exc, tb)
+            res, worker_id = loads(res_info)
             state['cache'][key] = res
             finish_task(dsk, key, state, results, keyorder.get)
             for f in posttask_cbs:
@@ -565,8 +563,7 @@ def get_sync(dsk, keys, **kwargs):
     Can be useful for debugging.
     """
     kwargs.pop('num_workers', None)    # if num_workers present, remove it
-    return get_async(apply_sync, 1, dsk, keys,
-                     raise_on_exception=True, **kwargs)
+    return get_async(apply_sync, 1, dsk, keys, **kwargs)
 
 
 def sortkey(item):
@@ -585,61 +582,3 @@ def sortkey(item):
     ('tuple', ('x', 1))
     """
     return (type(item).__name__, item)
-
-
-"""
-Remote Exceptions
------------------
-
-We want the following behaviors from remote exceptions
-
-1.  Include the original error message
-2.  Respond to try-except blocks with original error type
-3.  Include remote traceback
-"""
-
-
-class RemoteException(Exception):
-    """ Remote Exception
-
-    Contains the exception and traceback from a remotely run task
-    """
-    def __init__(self, exception, traceback):
-        self.exception = exception
-        self.traceback = traceback
-
-    def __str__(self):
-        return (str(self.exception) + "\n\n"
-                "Traceback\n"
-                "---------\n" +
-                self.traceback)
-
-    def __dir__(self):
-        return sorted(set(dir(type(self)) +
-                      list(self.__dict__) +
-                      dir(self.exception)))
-
-    def __getattr__(self, key):
-        try:
-            return object.__getattribute__(self, key)
-        except AttributeError:
-            return getattr(self.exception, key)
-
-
-exceptions = dict()
-
-
-def remote_exception(exc, tb):
-    """ Metaclass that wraps exception type in RemoteException """
-    if type(exc) in exceptions:
-        typ = exceptions[type(exc)]
-        return typ(exc, tb)
-    else:
-        try:
-            typ = type(exc.__class__.__name__,
-                       (RemoteException, type(exc)),
-                       {'exception_type': type(exc)})
-            exceptions[type(exc)] = typ
-            return typ(exc, tb)
-        except TypeError:
-            return exc

--- a/dask/compatibility.py
+++ b/dask/compatibility.py
@@ -54,6 +54,11 @@ if PY3:
     def _getargspec(func):
         return inspect.getfullargspec(func)
 
+    def reraise(exc, tb=None):
+        if exc.__traceback__ is not tb:
+            raise exc.with_traceback(tb)
+        raise exc
+
 else:
     import __builtin__ as builtins
     from Queue import Queue, Empty
@@ -72,6 +77,16 @@ else:
     reduce = reduce
     operator_div = operator.div
     FileNotFoundError = IOError
+
+    def _make_reraise():
+        _code = ("def reraise(exc, tb=None):"
+                "    raise type(exc), exc, tb")
+        namespace = {}
+        exec("exec _code in namespace")
+        return namespace['reraise']
+
+    reraise = _make_reraise()
+    del _make_reraise
 
     def _getargspec(func):
         return inspect.getargspec(func)

--- a/dask/dataframe/io/tests/test_parquet.py
+++ b/dask/dataframe/io/tests/test_parquet.py
@@ -353,7 +353,7 @@ def test_categories(fn):
     cats_set = ddf2.map_partitions(lambda x: x.y.cat.categories).compute()
     assert cats_set.tolist() == ['a', 'c', 'a', 'b']
     assert_eq(ddf.y, ddf2.y, check_names=False)
-    with pytest.raises(dask.async.RemoteException):
+    with pytest.raises(TypeError):
         # attempt to load as category that which is not so encoded
         ddf2 = dd.read_parquet(fn, categories=['x']).compute()
 

--- a/dask/multiprocessing.py
+++ b/dask/multiprocessing.py
@@ -1,6 +1,7 @@
 from __future__ import absolute_import, division, print_function
 
 import multiprocessing
+import traceback
 import pickle
 import sys
 
@@ -34,6 +35,94 @@ _loads = pickle.loads
 
 def _process_get_id():
     return multiprocessing.current_process().ident
+
+
+# -- Remote Exception Handling --
+# By default, tracebacks can't be serialized using pickle. However, the
+# `tblib` library can enable support for this. Since we don't mandate
+# that tblib is installed, we do the following:
+#
+# - If tblib is installed, use it to serialize the traceback and reraise
+#   in the scheduler process
+# - Otherwise, use a ``RemoteException`` class to contain a serialized
+#   version of the formatted traceback, which will then print in the
+#   scheduler process.
+#
+# To enable testing of the ``RemoteException`` class even when tblib is
+# installed, we don't wrap the class in the try block below
+class RemoteException(Exception):
+    """ Remote Exception
+
+    Contains the exception and traceback from a remotely run task
+    """
+    def __init__(self, exception, traceback):
+        self.exception = exception
+        self.traceback = traceback
+
+    def __str__(self):
+        return (str(self.exception) + "\n\n"
+                "Traceback\n"
+                "---------\n" +
+                self.traceback)
+
+    def __dir__(self):
+        return sorted(set(dir(type(self)) +
+                      list(self.__dict__) +
+                      dir(self.exception)))
+
+    def __getattr__(self, key):
+        try:
+            return object.__getattribute__(self, key)
+        except AttributeError:
+            return getattr(self.exception, key)
+
+
+exceptions = dict()
+
+
+def remote_exception(exc, tb):
+    """ Metaclass that wraps exception type in RemoteException """
+    if type(exc) in exceptions:
+        typ = exceptions[type(exc)]
+        return typ(exc, tb)
+    else:
+        try:
+            typ = type(exc.__class__.__name__,
+                       (RemoteException, type(exc)),
+                       {'exception_type': type(exc)})
+            exceptions[type(exc)] = typ
+            return typ(exc, tb)
+        except TypeError:
+            return exc
+
+
+try:
+    import tblib.pickling_support
+    tblib.pickling_support.install()
+    from dask.compatibility import reraise
+
+    def _pack_traceback(tb):
+        return tb
+
+except ImportError:
+    def _pack_traceback(tb):
+        return ''.join(traceback.format_tb(tb))
+
+    def reraise(exc, tb):
+        exc = remote_exception(exc, tb)
+        raise exc
+
+
+def pack_exception(e, dumps):
+    exc_type, exc_value, exc_traceback = sys.exc_info()
+    tb = _pack_traceback(exc_traceback)
+    try:
+        result = dumps((e, tb))
+    except BaseException as e:
+        exc_type, exc_value, exc_traceback = sys.exc_info()
+        tb = _pack_traceback(exc_traceback)
+        result = dumps((e, tb))
+    return result
 
 
 def get(dsk, keys, num_workers=None, func_loads=None, func_dumps=None,
@@ -83,8 +172,9 @@ def get(dsk, keys, num_workers=None, func_loads=None, func_dumps=None,
     try:
         # Run
         result = get_async(pool.apply_async, len(pool._pool), dsk3, keys,
-                           get_id=_process_get_id,
-                           dumps=dumps, loads=loads, **kwargs)
+                           get_id=_process_get_id, dumps=dumps, loads=loads,
+                           pack_exception=pack_exception,
+                           raise_exception=reraise, **kwargs)
     finally:
         if cleanup:
             pool.close()

--- a/dask/tests/test_async.py
+++ b/dask/tests/test_async.py
@@ -2,8 +2,7 @@ from __future__ import absolute_import, division, print_function
 
 import dask
 
-from dask.async import (start_state_from_dask, get_sync, finish_task, sortkey,
-                        remote_exception)
+from dask.async import start_state_from_dask, get_sync, finish_task, sortkey
 from dask.order import order
 from dask.utils_test import GetFunctionTestMixin, inc, add
 
@@ -174,22 +173,9 @@ def test_exceptions_propagate():
         assert False
     except MyException as e:
         assert "My Exception!" in str(e)
-        assert "Traceback" in str(e)
         assert 'a' in dir(e)
-        assert 'traceback' in dir(e)
-        assert e.exception.a == 1 and e.exception.b == 2
-        assert e.a == 1 and e.b == 2
-
-
-def test_remote_exception():
-    e = TypeError("hello")
-    a = remote_exception(e, 'traceback')
-    b = remote_exception(e, 'traceback')
-
-    assert type(a) == type(b)
-    assert isinstance(a, TypeError)
-    assert 'hello' in str(a)
-    assert 'traceback' in str(a)
+        assert e.a == 1
+        assert e.b == 2
 
 
 def test_ordering():

--- a/dask/tests/test_multiprocessing.py
+++ b/dask/tests/test_multiprocessing.py
@@ -9,7 +9,7 @@ import pytest
 
 from dask import compute, delayed
 from dask.context import set_options
-from dask.multiprocessing import get, _dumps, _loads
+from dask.multiprocessing import get, _dumps, _loads, remote_exception
 from dask.utils_test import inc
 
 
@@ -35,6 +35,18 @@ def test_errors_propagate():
     except Exception as e:
         assert isinstance(e, ValueError)
         assert "12345" in str(e)
+
+
+def test_remote_exception():
+    e = TypeError("hello")
+    a = remote_exception(e, 'traceback-body')
+    b = remote_exception(e, 'traceback-body')
+
+    assert type(a) == type(b)
+    assert isinstance(a, TypeError)
+    assert 'hello' in str(a)
+    assert 'Traceback' in str(a)
+    assert 'traceback-body' in str(a)
 
 
 def make_bad_result():

--- a/dask/threaded.py
+++ b/dask/threaded.py
@@ -5,6 +5,7 @@ See async.py
 """
 from __future__ import absolute_import, division, print_function
 
+import sys
 from collections import defaultdict
 from multiprocessing.pool import ThreadPool
 import threading
@@ -29,6 +30,10 @@ main_thread = current_thread()
 
 pools = defaultdict(dict)
 pools_lock = Lock()
+
+
+def pack_exception(e, dumps):
+    return e, sys.exc_info()[2]
 
 
 def get(dsk, result, cache=None, num_workers=None, **kwargs):
@@ -73,7 +78,7 @@ def get(dsk, result, cache=None, num_workers=None, **kwargs):
 
     results = get_async(pool.apply_async, len(pool._pool), dsk, result,
                         cache=cache, get_id=_thread_get_id,
-                        **kwargs)
+                        pack_exception=pack_exception, **kwargs)
 
     # Cleanup pools associated to dead threads
     with pools_lock:

